crates/oxc_allocator/src/clone_in.rs

```diff
@@ -0,0 +1,89 @@
+use crate::{Allocator, Box, Vec};
+
+/// A trait to explicitly clone an object into an arena allocator.
+///
+/// As a convention `Cloned` associated type should always be the same as `Self`,
+/// It'd only differ in the lifetime, Here's an example:
+///
+/// ```
+/// impl<'old_alloc, 'new_alloc> CloneIn<'new_alloc> for Struct<'old_alloc> {
+///     type Cloned = Struct<'new_alloc>;
+///     fn clone_in(&self, alloc: &'new_alloc Allocator) -> Self::Cloned {
+///         Struct { a: self.a.clone_in(alloc), b: self.b.clone_in(alloc) }
+///     }
+/// }
+/// ```
+///
+/// Implementations of this trait on non-allocated items usually short-circuit to `Clone::clone`;
+/// However, it **isn't** guaranteed.
+///
+pub trait CloneIn<'new_alloc>: Sized {
+    type Cloned;
+
+    fn clone_in(&self, alloc: &'new_alloc Allocator) -> Self::Cloned;
+}
+
+impl<'a, T, C> CloneIn<'a> for Option<T>
+where
+    T: CloneIn<'a, Cloned = C>,
+{
+    type Cloned = Option<C>;
+    fn clone_in(&self, alloc: &'a Allocator) -> Self::Cloned {
+        self.as_ref().map(|it| it.clone_in(alloc))
+    }
+}
+
+impl<'old_alloc, 'new_alloc, T, C> CloneIn<'new_alloc> for Box<'old_alloc, T>
+where
+    T: CloneIn<'new_alloc, Cloned = C>,
+{
+    type Cloned = Box<'new_alloc, C>;
+    fn clone_in(&self, alloc: &'new_alloc Allocator) -> Self::Cloned {
+        Self::Cloned::new_in(self.as_ref().clone_in(alloc), alloc)
+    }
+}
+
+impl<'old_alloc, 'new_alloc, T, C> CloneIn<'new_alloc> for Vec<'old_alloc, T>
+where
+    T: CloneIn<'new_alloc, Cloned = C>,
+{
+    type Cloned = Vec<'new_alloc, C>;
+    fn clone_in(&self, alloc: &'new_alloc Allocator) -> Self::Cloned {
+        Self::Cloned::from_iter_in(self.iter().map(|it| it.clone_in(alloc)), alloc)
+    }
+}
+
+impl<'a, T: Copy> CloneIn<'a> for std::cell::Cell<T> {
+    type Cloned = std::cell::Cell<T>;
+    fn clone_in(&self, _: &'a Allocator) -> Self::Cloned {
+        Self::Cloned::new(self.get())
+    }
+}
```

Comment:
I have a slight preference to import `std::cell::Cell` at top of the file, and use just `Cell` here, as it reduces noise a little. But that's a matter of taste - feel free to ignore me!